### PR TITLE
Demo to show the working of the incompatible_qos event callbacks.

### PR DESCRIPTION
**Pull request by jaisontj**
Related to [this](https://github.com/ros2/ros2/issues/822) feature request. The design and implementation details can also be found there.

Signed-off-by: Jaison Titus <jaisontj92@gmail.com>

### CODE REVIEW

File: quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+from rclpy.qos import QoSLivelinessPolicy
+from rclpy.qos import QoSProfile
+from rclpy.qos import QoSReliabilityPolicy
+from rclpy.qos_event import PublisherEventCallbacks
+from rclpy.qos_event import SubscriptionEventCallbacks
+
+
+def get_parser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
```
Comment (piraka9011) [inline quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: You can add choices to argparse: https://docs.python.org/3/library/argparse.html#choices

File: quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        qos_profile_publisher.liveliness_lease_duration = Duration(seconds=2)
+        qos_profile_subscription.liveliness_lease_duration = Duration(seconds=1)
+    elif qos_policy_name == 'reliability':
+        print('Reliability incompatibility selected.\n \
+Incompatibility condition: publisher reliability < subscripition reliability.\n \
+Setting publisher reliability to: BEST_EFFORT\n \
+Setting subscription reliability to: RELIABLE\n')
+        qos_profile_publisher.reliability = \
+            QoSReliabilityPolicy.RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT
+        qos_profile_subscription.reliability = \
```
Comment (piraka9011) [inline quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: Are the `\` needed here?

Comment (jaisontj) [reply quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: For long assignments it does look like it is needed, AFAIK. 

Another thing that can be done is to assign one of the variables a shorter name, like [this](https://stackoverflow.com/questions/29555608/break-a-long-assignment-into-two-lines-in-python)

And I think you can also put the R.H.S inside parenthesis.

File: quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py (hunk 1 above)
Comment (jaisontj) [reply quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: Done! This was helpful. Thank you :)

File: quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    qos_policy_name = parsed_args.incompatible_qos_policy_name
+    qos_profile_publisher = QoSProfile(depth=10)
+    qos_profile_subscription = QoSProfile(depth=10)
+
+    if qos_policy_name == 'durability':
+        print('Durability incompatibility selected.\n \
+Incompatibility condition: publisher durability kind < \
+subscripition durability kind.\n \
+Setting publisher durability to: VOLATILE\n \
+Setting subscription durability to: TRANSIENT_LOCAL\n')
```
Comment (ivanpauno) [inline quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: This is really hard to read, I would recommend this other style:
```python3
        print(
            'Durability incompatibility selected.\n'
            'Incompatibility condition: publisher durability kind < '
            ...
        )
```

Comment (ivanpauno) [reply quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: That comment applies in other places too.

Comment (mm318) [reply quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: Done.

**Comment by mm318** [comment]
@ros2/aws-oncall - please run this CI job
Gist: `https://gist.githubusercontent.com/mm318/576eae9c880a670ce32d73793fb78a91/raw/7274e2f29d97e2ddbec6805759a3de29e5d30c1b/ros2_qos_demos.repos`
BUILD args: `--packages-up-to quality_of_service_demo_cpp quality_of_service_demo_py`
TEST args: `--packages-select quality_of_service_demo_cpp quality_of_service_demo_py`
Job: `ci_launcher`

File: quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py (hunk 3 above)
Comment (ivanpauno) [reply quality_of_service_demo/rclpy/quality_of_service_demo_py/incompatible_qos.py]: Thanks!

**Comment by dabonnie** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9874/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5492/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8034/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9732/)

**Comment by mm318** [comment]
@ivanpauno, there is a random buildfarm failure for the Linux-aarch64 run. Can we merge this pull request?

**Comment by ivanpauno** [comment]
> @ivanpauno, there is a random buildfarm failure for the Linux-aarch64 run. Can we merge this pull request?

Yes, I think it's safe to merge (considering that the last commit are just style changes and this PR has already passed CI before).
Thanks for the contribution!

**Pull request merged**